tests/PrefixCacheBootstrapperTest.php

```diff
@@ -0,0 +1,345 @@
+});
+
+test('Tenancy overrides CacheManager', function() {
+    expect(app('cache')::class)->toBe(TenancyCacheManager::class);
+    expect(app(CacheManager::class)::class)->toBe(TenancyCacheManager::class);
+
+    tenancy()->initialize(Tenant::create(['id' => 'first']));
+
+    expect(app('cache')::class)->toBe(TenancyCacheManager::class);
+    expect(app(CacheManager::class)::class)->toBe(TenancyCacheManager::class);
+
+    tenancy()->initialize(Tenant::create(['id' => 'second']));
+
+    expect(app('cache')::class)->toBe(TenancyCacheManager::class);
+    expect(app(CacheManager::class)::class)->toBe(TenancyCacheManager::class);
+
+    tenancy()->end();
+
+    expect(app('cache')::class)->toBe(TenancyCacheManager::class);
+    expect(app(CacheManager::class)::class)->toBe(TenancyCacheManager::class);
+});
+
+test('correct cache prefix is used in all contexts', function () {
+    $originalPrefix = config('cache.prefix');
+    $prefixBase = config('tenancy.cache.prefix_base');
+    $expectPrefixToBe = function(string $prefix) {
+        expect($prefix . ':') // RedisStore suffixes prefix with ':'
+            ->toBe(app('cache')->getPrefix())
+            ->toBe(app('cache.store')->getPrefix());
+    };
+
+    $expectPrefixToBe($originalPrefix);
+
+    $tenant1 = Tenant::create();
+    $tenant2 = Tenant::create();
+
+    $tenantOnePrefix = $originalPrefix . $prefixBase . $tenant1->getTenantKey();
+
+    tenancy()->initialize($tenant1);
+    // Default prefix generator
+    $prefixGenerator = PrefixCacheTenancyBootstrapper::$prefixGenerator;
+
+    cache()->set('key', 'tenantone-value');
+
+    $expectPrefixToBe($prefixGenerator($tenant1));
+
+    $tenantTwoPrefix = $originalPrefix . $prefixBase . $tenant2->getTenantKey();
+
+    tenancy()->initialize($tenant2);
+
+    cache()->set('key', 'tenanttwo-value');
+
+    $expectPrefixToBe($prefixGenerator($tenant2));
+
+    // Prefix gets reverted to default after ending tenancy
+    tenancy()->end();
+    $expectPrefixToBe($originalPrefix);
+
+    // Assert tenant's data is accessible using the prefix from the central context
+    config(['cache.prefix' => null]); // stop prefixing cache keys in central so we can provide prefix manually
+    app('cache')->forgetDriver(config('cache.default'));
+
+    expect(cache($tenantOnePrefix . ':key'))->toBe('tenantone-value');
+    expect(cache($tenantTwoPrefix . ':key'))->toBe('tenanttwo-value');
+});
+
+test('cache is persisted when reidentification is used', function () {
+    $tenant1 = Tenant::create();
+    $tenant2 = Tenant::create();
+    tenancy()->initialize($tenant1);
+
+    cache(['foo' => 'bar'], 10);
+    expect(cache('foo'))->toBe('bar');
+
+    tenancy()->initialize($tenant2);
+    tenancy()->end();
+
+    tenancy()->initialize($tenant1);
+    expect(cache('foo'))->toBe('bar');
+});
+
+test('prefixing separates the cache', function () {
+    $tenant1 = Tenant::create();
+    tenancy()->initialize($tenant1);
+
+    cache()->put('foo', 'bar', 1);
```

Comment:
What's the `1` for?

Reply:
Good question, it's TTL, but I don't see a reason for setting TTL here (also in some other `cache()->put()` calls in the tests)

Reply:
I deleted that